### ISSUE TITLE
LargePrecisionTensor

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#2145]
In many cases, native hardware does not support precision which is large enough for the cryptographic use cases we'd like to solve. For example, Android cannot support larger types than Float (or unsigned Int), and PyTorch cannot support larger types than a Signed Long. Unfortunately, we need types that are much larger than this (256 bytes in some cases would be ideal). 

In this project, we would like to have a new tensor type which can support arbitrarily large values by simply packing them into **multiple smaller values**. For example, a 2x5 tensor with 128 Bit Integers, might be backed by an IntTensor of dimension 2x5x4, where the last dimension (4 ints) is interpreted to be a single long integer.

This will require overloading relevant arithmetic API calls (such as addition, multiplication, etc.) to properly interpret data stored in this format.

**Comment by mccorby** [comment]
I'll pick this up

**Comment by mccorby** [comment]
What should we do when two `LargePrecisionTensor`s represent numbers of different precision?
For example, let n1 in 2**128 and n2 in 2**64 and objective bits 32
This will result in two matrices with different dimensions: matrix_n1 1x4 and matrix_n2 1x2. 

I can find a way of doing (finding the offset and then only taking into account relevant cells but it will be cumbersome and slow). So if anyone can give me a hint, more than welcomed!

**Comment by Jasopaum** [comment]
Could padding be a solution in your case? I mean casting n2 in 2^128 with all the MSB at 0

**Comment by mccorby** [comment]
I thought about it but then what to do in a multiplication?

**Comment by LaRiffle** [comment]
Maybe you could start by enforcing similar precision just as we might have done for FixedPrecision tensors?

**Comment by mccorby** [comment]
I've been doing some progress on this issue (see https://github.com/OpenMined/PySyft/pull/2147/) but I'm not sure I am going in the right direction.
There are a few things I'm finding for which I'd like some input:
- Am I overcomplicated things :) ?
- A `torch.Tensor` can't be at the head in the chain. If this would be possible we wouldn't need this class!
- I'm using `.child` to hold the tensor generated from the original array. Not sure if this is correct
- The input is now a `np.array`. Numpy arrays can hold very big numbers as `object`s
- Numbers of different sizes are decomposed into tensors of different dimensions. There is a function that fill up the number with the minor last dimension so that the tensors can be operated with. 
- Multiplication is still to be done and I'm trying to figure out how to do it

It'd be good to have a scenario where these tensors could be put to work. If anyone can provide with one I would be happy to make it work (if all the above is still valid!)

Cheers

**Comment by LaRiffle** [comment]
Hey @mccorby I'm looking into it right now,
I think you're doing the right thing no worries :)
So actually it would be in a chain where indeed the head would be a `torch.Tensor`, but maybe in this kind of way: I have torch.tensor([3.0]) which I want to express in fixed precision with fixed_precision=512 (stupid but why not). In that case, the head is that torch.tensor and below I get the LargePrecisionTensor

**Comment by LaRiffle** [comment]
Another point: maybe first start with operate together tensors of the same dimension, which is determined by the precision parameter. This will give you tensors fo the same size which is helpful. This size will be determined by the values allocated for the decimal numbers (directly linked the precision parameter) and the values of the integers part, which you can first say it's only the size of a single LongTensor, and we'll add a integer precision parameter for it as well later, as we used to have before fore fixed precision tensors (https://github.com/OpenMined/PySyft/commit/32dc45160a7bce78dc1dbe750974b548eba26117#diff-eee9ccaf3c5a0ebc4cf135d97bbd3fbb)

**Comment by mccorby** [comment]
I created a draft with one the possibilities I see we could use to deal with the operations of large numbers. Basically it uses NumPy for such operations. 
The main difference with the other PR is that the number is reconstructed before operating (see `def add()`.

https://github.com/OpenMined/PySyft/pull/2257/files

Please note this is not a PR to merge but only for discussing it

**Comment by Jasopaum** [comment]
I am not sure using numpy solves the original problem stated above: "native hardware does not support precision which is large enough for the cryptographic use cases we'd like to solve".
But maybe I misunderstood something!

**Comment by mccorby** [comment]
NumPy works with `objects` to represent large numbers. Not sure how it operates with them internally but it seems to be working.
If we're not using NumPy, we need then to find out how to solve the issue with the overflow and how to do the multiplications.
One problem (or another) of using NumPy is that we'll need to find an equivalent in other languages. There is a promising one for the JVM but I still need to try it (KMath).

Maybe what we need is a new way of computing based only on big integers :)

**Comment by mccorby** [comment]
```

**Comment by mccorby** [comment]
> But maybe I misunderstood something!
Maybe it's me who is misunderstanding everything :)

**Comment by Jasopaum** [comment]
Oh ok, so all the computations are done with numpy that handles large number on its own even on constrained hardware! Sorry, I was not seeing things like that 
So you might also need to use a np object in your _restore_large_number function then, don't you?

**Comment by mccorby** [comment]
If we go for this solution there are a few other things to fix too.
